internal/auth/ldap/repository_auth_method_update.go

```diff
@@ -0,0 +1,563 @@
+	}
+	if err := validateFieldMask(ctx, fieldMaskPaths); err != nil {
+		return nil, db.NoRowsAffected, errors.Wrap(ctx, err, op)
+	}
+	dbMask, nullFields := dbw.BuildUpdatePaths(
+		map[string]any{
+			NameField:                 am.Name,
+			DescriptionField:          am.Description,
+			StartTlsField:             am.StartTls,
+			InsecureTlsField:          am.InsecureTls,
+			DiscoverDnField:           am.DiscoverDn,
+			AnonGroupSearchField:      am.AnonGroupSearch,
+			UpnDomainField:            am.UpnDomain,
+			UserDnField:               am.UserDn,
+			UserAttrField:             am.UserAttr,
+			UserFilterField:           am.UserFilter,
+			GroupDnField:              am.GroupDn,
+			GroupAttrField:            am.GroupAttr,
+			GroupFilterField:          am.GroupFilter,
+			CertificatesField:         am.Certificates,
+			ClientCertificateField:    am.ClientCertificate,
+			ClientCertificateKeyField: am.ClientCertificateKey,
+			BindDnField:               am.BindDn,
+			BindPasswordField:         am.BindPassword,
+			UrlsField:                 am.Urls,
+		},
+		fieldMaskPaths,
+		nil,
+	)
+	if len(dbMask) == 0 && len(nullFields) == 0 {
+		return nil, db.NoRowsAffected, errors.New(ctx, errors.EmptyFieldMask, op, "empty field mask")
+	}
+	if strutil.StrListContains(nullFields, UrlsField) {
+		return nil, db.NoRowsAffected, errors.New(ctx, errors.InvalidParameter, op, "missing urls (you cannot delete all of them; there must be at least one)")
+	}
+
+	origAm, err := r.LookupAuthMethod(ctx, am.PublicId)
+	if err != nil {
+		return nil, db.NoRowsAffected, errors.Wrap(ctx, err, op)
+	}
+	if origAm == nil {
+		return nil, db.NoRowsAffected, errors.New(ctx, errors.RecordNotFound, op, fmt.Sprintf("auth method %q", am.PublicId))
+	}
+	// there's no reason to continue if another controller has already updated this auth method.
+	if origAm.Version != version {
+		return nil, db.NoRowsAffected, errors.New(ctx, errors.VersionMismatch, op, fmt.Sprintf("update version %d doesn't match db version %d", version, origAm.Version))
+	}
+
+	dbWrapper, err := r.kms.GetWrapper(ctx, origAm.ScopeId, kms.KeyPurposeDatabase)
+	if err != nil {
+		return nil, db.NoRowsAffected, errors.Wrap(ctx, err, op, errors.WithMsg("unable to get database wrapper"))
+	}
+	addUrls, deleteUrls, err := valueObjectChanges(ctx, origAm.PublicId, UrlVO, am.Urls, origAm.Urls, dbMask, nullFields)
+	if err != nil {
+		return nil, db.NoRowsAffected, errors.Wrap(ctx, err, op)
+	}
+	addCerts, deleteCerts, err := valueObjectChanges(ctx, origAm.PublicId, CertificateVO, am.Certificates, origAm.Certificates, dbMask, nullFields)
+	if err != nil {
+		return nil, db.NoRowsAffected, errors.Wrap(ctx, err, op)
+	}
+	var addUserSearchConf, deleteUserSearchConf any
+	if strutil.StrListContains(dbMask, UserDnField) || strutil.StrListContains(dbMask, UserAttrField) || strutil.StrListContains(dbMask, UserFilterField) {
```

Comment:
This function does a linear search every time. Maybe this logic could benefit from a `StrListContainsAny(list []string, needles ...[]string)`? It could create a set of needles for O(1) lookup if there are enough needles.

Reply:
There's also https://pkg.go.dev/golang.org/x/exp/slices now which has a `Contains` method.

Reply:
I'll wait for the exp pkg to move to the stdlib, but I'm looking fwd to it.

Reply:
pushed a fixup, better?